src/main/scala/async/Listener.scala

```diff
@@ -1,73 +1,171 @@
 package gears.async
 
-import gears.async.Listener.Locked
-import gears.async.Listener.Failed
-import gears.async.Listener.SemiLock
 import scala.annotation.tailrec
 import gears.async.Async.Source
+import java.util.concurrent.locks.ReentrantLock
 
 object Listener:
+  /** The result of locking a single listener lock. */
   sealed trait LockResult
 
-  object Locked extends LockResult
+  /** We have completed locking the listener. It can now be `complete`d. */
+  case object Locked extends LockResult
 
-  object Failed extends LockResult
+  /** The listener is no longer available.
+    * It should be removed from the source, and any acquired locks should be `release`d.
```

Comment:
```suggestion
    * The lock of the returning stage itself is released but upstream locks have to be released by the requesting source.
    * The source can then assume this listener will never succeed again and drop it.
```

Reply:
Hm, right now the way `race` was implemented is that when `lockNext` fails the `PartialLock` is still held, and so locks are to be released up to the given partial lock.

Reply:
Sources should drop the failing listener it called `lockSelf` on. Listeners can (should) drop itself from other subscribed sources within `lockSelf`, if possible.